aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -816,6 +841,31 @@ SparseTensor sparse_mask(const Tensor& t, const SparseTensor& mask) {
   return t.mul(mask_template).to(t.scalar_type());
 }
 
+Tensor sparse_mask_projection(const Tensor& t, const Tensor& mask) {
+  TORCH_INTERNAL_ASSERT(t.is_sparse());
+  TORCH_INTERNAL_ASSERT(mask.is_sparse());
+
+  TORCH_CHECK(
+      mask.sizes().equals(t.sizes()),
+      "sparse_mask(): operands have incompatible sizes; self has size ",
+      t.sizes(),
+      " but mask has size ",
+      mask.sizes());
+
+  if (!t.numel() || !t._nnz() || !mask.numel() || !mask._nnz()) {
+    auto res = t.clone();
```

Comment:
Shouldn't `mask` be used instead of `t`? This is because the returned tensor should have the same indices as the `mask` has.

Reply:
So, do we allow the indices of `sparse_mask_like_grad` (and consequently `sparse_sparse_matmul_backward`) result to be variable: sometimes these are the indices of `x`, and sometimes the indices of `gx`?

Reply:
Yes, depending on what is projected onto what. To eliminated syncs, it is beneficial to project a sorted tensor onto a coalesced one.

Reply:
> No, it has different semantics.

What are the semantics differences between `_sparse_mask_projection` and `sparse_mask`?

Reply:
Projection is "reversed" and matches are not accumulated, as to not produce gradients which are multiples of the in-flowing gradients.

Reply:
@nikitaved this description is a little cryptic for me. Could explain the difference in terms of operation invariants rather than in the terms of a need (for which I have no doubts)?

For example, in the case of `sparse_mask(self, mask)`, we [have](https://pytorch.org/docs/main/generated/torch.Tensor.sparse_mask.html?highlight=sparse_mask#torch.Tensor.sparse_mask):
> Returns a new sparse tensor with values from a strided tensor `self` filtered by the indices of the sparse tensor mask. The values of `mask` sparse tensor are ignored. `self` and `mask` tensors must have the same shape.
> The returned sparse tensor might contain duplicate values if `mask` is not coalesced.
> The returned sparse tensor has the same indices as the sparse tensor `mask`.

How would you rephrase this documentation bit for `_sparse_mask_projection(mask)`?

Reply:
`self._sparse_mask_projection(mask)` will a tensor with the same indices as `self` and values which are nonzero for all matches between `self.indices()` and `mask.indices()`. The word "filter" is a bit confusing, as it does not specify the behavior of `sparse_mask`-like functions with uncoalesced inputs.

Reply:
Thanks, @nikitaved! So, IIUC, we have the following differences in-between `sparse_mask` and `_sparse_mask_projection`:
1. `sparse_mask` `input` and `mask` can have any layouts (`strided`, `sparse_coo`, `sparse_csr`, ... but `mask` must have some sparse layout), while `_sparse_mask_projection` supports only `sparse_coo` layout both in `input` and `mask`
2. `sparse_mask` result indices are the same as `mask` indices, while `_sparse_mask_projection` indices are the same as `input` indices

In other words, if `y = x.sparse_mask(mask)` and `z = x._sparse_mask_projection(mask)` then `z = x.sparse_mask(mask.sparse_mask(x))`. Is this correct?

Reply:
Agree with 1 and 2. The last point I disagree with (a mistake there?). These methods are never equivalent in general if at least one of the inputs is uncoalesced.

Reply:
If only `sparse_mask` did not require indices to be taken from `mask`, I would argue that `_sparse_mask_projection` is a better version of `sparse_mask` for sparse COO inputs, as it just does the filtering (match ? include : set to zero).

Reply:
OK, I see. `sparse_mask` has the weirdness of duplicating values when the mask is uncoalesced. So, something like the following should then hold:
`x._sparse_mask_projection(mask).coalesce()` is the same as `x.sparse_mask(mask.sparse_mask(x.coalesce()))`

Reply:
Yes, although `sparse_mask` sums up the duplicates, unlike `_sparse_mask_projection`. In this case we can for sure that as long as `mask` is coalesced, so is `res.sparse_mask(mask)` no matter whether `res` is coalesced or not.

Reply:
It is a doubling only if there are exactly two matches. A duplication would mean several memory copies and we do not have these.